Add Prometheus metrics endpoint

neuralsandwich opened the pull request:
## Description

Add basic Prometheus metrics to OAuth2 Proxy. Allow collected metrics to be scraped from "/metrics".

The following metrics have been added:
* oauth2_proxy_requests_total
* oauth2_proxy_requests_in_flight
* oauth2_proxy_responce_duration_seconds

## Motivation and Context

This is to address #580, Prometheus is used in many environments to track an applications behaviours and help diagnose issues.

## How Has This Been Tested?

* All existing tests were run successfully
* Metrics middleware test were added
  * These are basic tests but prove the functionality
* OAuth2 Proxy was configured and run with Okta with metrics being scraped via Prometheus
  * User login was tested
  * Unauthorized user was tested

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

NickMeves commented:
Nice! Thanks for picking this up! I'll give it a thorough look over the next few days.

Quick note from eyeballing the diff:

To allow users of OAuth2-Proxy to layer on as many supplemental security controls on top of this security tool, I'm gonna be pretty steadfast in my request that any metrics endpoint be hosted on its own metrics port. That way, users can layer on much more secure port based firewalls easily if desired.

A lot of installations have OAuth2-Proxy public facing or as a BeyondCorp style access proxy, `/metrics` shouldn't be exposed openly on `4180` leaking operational metadata.

neuralsandwich commented:
Yeah, I had a similar thought of requiring the agent contacting the endpoint to be authenticated. The simplest solution for me was to present a simple solution and discuss it from there :)

JoelSpeed reviewed (commented):
This is great! Thanks for making a start on adding the metrics

I added a bunch of suggestions inline, please ping me if you need any more explanation on any of them

I second the fact these should be on a separate port, this would make it much easier for people to lock down their metrics and is a common pattern across other projects. We should be able to use very basic http server for now to get the second metrics port up and serving

neuralsandwich commented on pkg/apis/options/options.go:
```diff
@@ -207,6 +209,7 @@ func NewFlagSet() *pflag.FlagSet {
 	flagSet.String("proxy-prefix", "/oauth2", "the url root path that this proxy should be nested under (e.g. /<oauth2>/sign_in)")
 	flagSet.String("ping-path", "/ping", "the ping endpoint that can be used for basic health checks")
 	flagSet.String("ping-user-agent", "", "special User-Agent that will be used for basic health checks")
+	flagSet.String("metrics-path", "/metrics", "the metrics endpoint that can be used for scraping metrics")
```
👍 I started off copying the healthcheck implementation for figuring out how to get middleware working.

neuralsandwich commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,120 @@
+	metricsPath string) alice.Constructor {
+	return func(next http.Handler) http.Handler {
+
+		// Counter for all requests
+		// This is bucketed based on the response code we set
+		requestsCounter := RegisterRequestsCounter(registerer)
+
+		// Gauge to all requests currently being handled
+		inFlightRequests := RegisterInflightRequestsGauge(registerer)
+
+		// The latency of all requests bucketed by HTTP method
+		requestLatency := RegisterRequestsLatencyHistogram(registerer)
```
Yeah, it was mostly for testing.

JoelSpeed replied:
It may come in useful in the future. Using the global registry all sorts of other packages will add their own metrics, that may or may not be desirable so we may want to use a custom registry

neuralsandwich commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,120 @@
+		[]string{"code"},
+	)
+
+	// Initialize likely HTTP status codes.
```
These are all the code I could see in the codebase. At least those using `http.Status*`.

JoelSpeed replied:
Thanks for checking! That satisfies my concern :)

When you say they might be... Are there any docs or anything we could refer to or is it a case we need to dig through the code and try to work it out?

neuralsandwich commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,120 @@
+	return histogram
+}
+
+func MetricsServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	promhttp.Handler().ServeHTTP(rw, req)
+}
```
It will display the formatted metrics but it would only display metrics register with the default registry. So yeah, missing our custom metrics but currently it is only using default registry. Without storing the registry somewhere and passing it here we wouldn't be able to display them.

Since we are moving this to a different port any how, I'll push some things into `oauthproxy.go` and wire a handler for the custom registry.

JoelSpeed replied:
I might suggest doing as much as you can in a `metrics.go` file rather than adding to the huge `oauthproxy.go` file

neuralsandwich commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,120 @@
+	return NewRequestInstrumentor(prometheus.DefaultRegisterer, metricsPath)
+}
+
+func NewRequestInstrumentor(registerer prometheus.Registerer,
```
I could do it if we only want the default version to be used. Otherwise, I could rename the WithDefaultRegistry version to this.

neuralsandwich commented:
I've updated the PR based on the feedback. I haven't made `NewRequestInstrumentor` private but can do or I could rename `NewRequestInstrumentorWithDefaultRegistery` to `NewRequestInstrumentor`.

I've also added a new HTTP server to `main.go` listening to 9100 by default.

neuralsandwich commented:
I've collapsed a few calls in main to `Printf` and then `os.Exit` to `Fatalf` since they are equivalent to get around the hard 50 lines limit in main.

JoelSpeed commented on main.go:
```diff
@@ -50,19 +51,20 @@ func main() {
 	rand.Seed(time.Now().UnixNano())
 
+	http.HandleFunc("/metrics", middleware.MetricsServe)
+	go http.ListenAndServe(opts.MetricsPort, nil)
```
Is there a way to have this stop gracefully as we do in the main server?

NickMeves replied:
Should we also have this be opt-in and disabled from this metrics server port being turned on by default?

For users who don't have prometheus in their environment, it would be best to not open up extra unneeded ports from a security perspective.

neuralsandwich replied:
@JoelSpeed I can look into making this more graceful.

@NickMeves We could if that is what is desired. With this being on a different port, it shouldn't be exposed if the user has correctly set up their firewall/docker ports. What should be the default mode, metrics on by default or off?

JoelSpeed replied:
> @JoelSpeed I can look into making this more graceful.

This could be left as a TODO if it's a lot of work at this point, I'll let you make this call

> What should be the default mode, metrics on by default or off?

Off by default IMO, perhaps a special value of `0` makes it off and is the default value?

neuralsandwich replied:
":0" technically means bind to a random free port in Windows and Linux. I could change the default to an empty string and only enable the server when we are provided an address. Also might be worth changing the flag to address instead of port as well.

NickMeves replied:
I lean toward a dedicated flag on whether metrics is enabled or disabled. Then a separate flag specifying the port to listen on if enabled.

JoelSpeed replied:
Perhaps we should change this to `MetricsBindAddress` and accept a host and port, then have empty string as off, eg I could set `127.0.0.1:8080` for only binding to localhost, is that relatively easy to achieve?

neuralsandwich replied:
I didn't see this comment but I have essentially implemented this I can rename the function.

JoelSpeed commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,109 @@
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var metricsHandler = NewInstrumentMetricHandlerWithDefaultRegistry()
```
Instead, what about this?
```suggestion
var ServeMetrics = NewInstrumentMetricHandlerWithDefaultRegistry().ServeHTTP
```
Then in main where you have `middleware.MetricsServe` you replace it with `middleware.ServeMetrics`? (I prefer the name this way round)

Then we don't need the function wrapper at the end of this file

neuralsandwich replied:
Yeah, that makes more sense.

neuralsandwich replied:
I've stuck with MetricsHandler for now due to the change in handling shutdown in `main.go`. I'm happy to change if you think it should still be named differently.

JoelSpeed commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,109 @@
+
+var metricsHandler = NewInstrumentMetricHandlerWithDefaultRegistry()
+
+func NewInstrumentMetricHandlerWithDefaultRegistry() http.Handler {
```
Would be nice to have comments on each of the functions in this file as we do in others in this package

JoelSpeed commented on pkg/middleware/instrumentation_test.go:
```diff
@@ -0,0 +1,67 @@
+			err = testutil.GatherAndCompare(in.registry, expectedPrometheusText, in.expectedMetrics...)
+			Expect(err).NotTo(HaveOccurred())
+		},
+		Entry("successfully", func() *requestTableInput {
+			in := &requestTableInput{
+				registry:      prometheus.NewRegistry(),
+				requestString: "http://example.com/metrics",
+				expectedMetrics: []string{
+					"oauth2_proxy_requests_total",
+				},
+				expectedStatus:      200,
+				expectedResultsFile: "testdata/metrics/successfulrequest.txt",
+			}
+			in.expectedHandler = NewInstrumentMetricHandler(in.registry, in.registry)
+
+			return in
+		}()),
```
Well that's an interesting pattern I haven't seen before! Seems to work so I'm not complaining, just took me a moment to work out what was happening 😅

neuralsandwich replied:
I attempted to copy the structure used in `healthcheck_test.go`. Happy to change anything to make it clearer or easier to maintain.

JoelSpeed replied:
I think this is fine, no worries, no need to change!

JoelSpeed reviewed (commented):
Couple of suggestions, but otherwise I'm happy with this, nothing too major really

Thanks for working with us on this one!

NickMeves commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,109 @@
+package middleware
```
NIT: Can we name this middleware file to more accurately reflect it is for prometheus/metrics? `instrumentation.go` is vague in a vacuum when in broader `middleware` package and not a more focused `prometheus` package where that term makes more sense in context.

neuralsandwich replied:
Can do.

NickMeves commented on oauthproxy.go:
```diff
@@ -256,6 +256,8 @@ func buildPreAuthChain(opts *options.Options) (alice.Chain, error) {
 		chain = chain.Append(LoggingHandler, middleware.NewHealthCheck(healthCheckPaths, healthCheckUserAgents))
 	}
 
+	chain = chain.Append(middleware.NewRequestInstrumentorWithDefaultRegistry())
```
Same NIT as the file name -- should this middleware handler have naming that reflects it is metrics/prometheus based?

Also, this method name is a bit of a mouthful 😅

NickMeves replied:
Question: with this being a global middleware, do we not have any of our instrumented promhttp metrics broken down per endpoint?

For instance, health checks vs backend proxy requests are very different and I don't think either of those metrics would want the other polluting their dataset.

NickMeves replied:
Related thought on that front (scope creep; don't worry about it here) -- there might be good metrics for us to collect as a different set of middleware attached to the session authentication middleware chain in addition to these metrics we are grabbing from the request middleware chain.

neuralsandwich replied:
Yeah, I think there is follow up work to do here to figure out what metrics are important to track. My goal for the PR was to get _some_ metrics going, get feedback and improve :D

I'll work on the name and try to make it clearer that this is going to capture metrics data for Prometheus... I'm definitely not the best at naming :joy:

NickMeves commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,109 @@
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			gauge = are.ExistingCollector.(prometheus.Gauge)
+		} else {
+			panic(err)
```
Is `panic` correct in these scenarios? Or should we control the `logger.Errorf` or `logger.Fatalf` in a more structured manner before we either panic or exit?

NickMeves replied:
Oh nevermind, I think I see the documentation on how prometheus behaves if the wrapped handler panics.

NickMeves commented on pkg/middleware/instrumentation.go:
```diff
@@ -0,0 +1,109 @@
+			return promhttp.InstrumentHandlerDuration(registerRequestsLatencyHistogram(registerer), next)
+		}
+
+		return alice.New(counterHandler, inFlightHandler, durationHandler).Then(next)
```
These definitely seem like the most important toward getting good RED metrics 👍 

`InstrumentHandlerResponseSize` would be another one that might be worth us instrumenting in the future. (the RequestSize variant too maybe?)

NickMeves commented on pkg/apis/options/options.go:
```diff
@@ -22,6 +22,7 @@ type Options struct {
 	ProxyPrefix        string   `flag:"proxy-prefix" cfg:"proxy_prefix"`
 	PingPath           string   `flag:"ping-path" cfg:"ping_path"`
 	PingUserAgent      string   `flag:"ping-user-agent" cfg:"ping_user_agent"`
+	MetricsPort        string   `flag:"metrics-port" cfg:"metrics_port"`
```
Once we get to the finish line on the code content aspect of this PR, we should be sure to get these new flags into the documentation. Prometheus support and which metrics we are exposing likely deserve there own documentation page explaining everything too.

neuralsandwich replied:
Added the flags to the docs. Might not be all the right places so I am happy to add more if you could point me to the right place.

JoelSpeed commented on main.go:
```diff
@@ -48,21 +50,21 @@ func main() {
 	rand.Seed(time.Now().UnixNano())
 
+	m := &http.Server{Addr: opts.MetricsPort, Handler: middleware.DefaultMetricsHandler}
+	m.ListenAndServe()
```
IIRC this is a blocking call, it will need to be in a go routine right?

neuralsandwich replied:
Yes, it will, oops. :facepalm:

JoelSpeed commented on docs/docs/features/endpoints.md:
```diff
@@ -7,6 +7,7 @@ OAuth2 Proxy responds directly to the following endpoints. All other endpoints w
 
 - /robots.txt - returns a 200 OK response that disallows all User-agents from all paths; see [robotstxt.org](http://www.robotstxt.org/) for more info
 - /ping - returns a 200 OK response, which is intended for use with health checks
+- /metrics - Metrics endpoint for Prometheus to scrape, serve on the address specified by `--metrics-address`, 127.0.0.1:9100 by default
```
Disabled by default now right?
```suggestion
- /metrics - Metrics endpoint for Prometheus to scrape, serve on the address specified by `--metrics-address`, disabled by default
```

JoelSpeed commented on docs/docs/configuration/overview.md:
```diff
@@ -89,6 +89,7 @@ An example [oauth2-proxy.cfg](https://github.com/oauth2-proxy/oauth2-proxy/blob/
 | `--provider-display-name` | string | Override the provider's name with the given string; used for the sign-in page | (depends on provider) |
 | `--ping-path` | string | the ping endpoint that can be used for basic health checks | `"/ping"` |
 | `--ping-user-agent` | string | a User-Agent that can be used for basic health checks | `""` (don't check user agent) |
+| `--metrics-address` | string | the address prometheus metrics will be scraped from | `""` |
```
Might be good to have an example or two of the expected format, WDYT?
```suggestion
| `--metrics-address` | string | the address prometheus metrics will be scraped from (e.g. `127.0.0.1:9100`, `:9100`) | `""` |
```

JoelSpeed commented on main.go:
```diff
@@ -73,11 +74,36 @@ func main() {
+	// Attempt to setup the metrics endpoint if we have an address
+	if address != "" {
+		s := &http.Server{Addr: address, Handler: middleware.DefaultMetricsHandler}
+		go s.ListenAndServe()
```
What happens if there is an error starting this up, eg and invalidly formatted address?

neuralsandwich replied:
Good point. I have wrapped this in `logger.Fatal()` now but I'm not sure if that is the right call. If the metrics address is invalid should it fail to start or just fail to run the metrics and display an error?

JoelSpeed commented on main.go:
```diff
@@ -73,11 +74,36 @@ func main() {
+				logger.Fatal(err)
+			}
+		}()
+	} else { // Capture the stop signal as not block shutdown
+		go func() {
+			<-stop
+		}()
```
I wonder if instead of writing to these channels, we need to close them instead? That would mean we don't need to have this block. Reading from a closed channel doesn't block so works in the same way as writing to a channel, but you don't have anything read from it

neuralsandwich replied:
I've removed this and now close the channel instead. The metrics HTTP server still closes so I think this was redundant. Thanks for catching that.

JoelSpeed commented on pkg/apis/options/options.go:
```diff
@@ -207,6 +209,7 @@ func NewFlagSet() *pflag.FlagSet {
 	flagSet.String("proxy-prefix", "/oauth2", "the url root path that this proxy should be nested under (e.g. /<oauth2>/sign_in)")
 	flagSet.String("ping-path", "/ping", "the ping endpoint that can be used for basic health checks")
 	flagSet.String("ping-user-agent", "", "special User-Agent that will be used for basic health checks")
+	flagSet.String("metrics-address", "", "the address /metrics will be served on")
```
Would be good to add the examples here too

JoelSpeed reviewed (commented):
No worries about the rename, this is looking great! Left a few comments but nothing really major.

I think the way the shutdown is being handled right now is probably ok, but let me know what you think of my suggestion

neuralsandwich commented:
I should be able to address the feedback on Monday. Busy week, just wanted to update to say this hasn't been abandoned.

JoelSpeed commented on main.go:
```diff
@@ -73,11 +74,35 @@ func main() {
+	if address != "" {
+		s := &http.Server{Addr: address, Handler: middleware.DefaultMetricsHandler}
+		go func() {
+			logger.Fatal(s.ListenAndServe())
```
Only need to log this if there is an error, or does logger.Fatal check if the error is nil first? 🤔  This is more explicit either way

```suggestion
			if err := s.ListenAndServe(); err != nil {
				logger.Fatal(err)
			}
```

As for whether we should quit if the metrics don't start, I would suggest that's reasonable behaviour, I'd expect if I had requested metrics, I'd want to know if they weren't working

neuralsandwich replied:
`ListenAndServe()` always returns `ErrServerClosed` so there is no need to check for an error. `logger.Fatal(s.ListenAndServe())` is the example used in the `http` package and seems like appropriate usage here. If `ListenAndServe()` stops, our metrics would be unavailable and we wouldn't want to continue running (and if the metrics were monitor we might be signalled to stop anyway).

That being said, we could call `logger.Print` here instead and send a message to the main HTTP server's stop channel. Calling `os.Exit()` as `Fatal` may not be the most graceful approach in retrospect.

neuralsandwich commented on main.go:
```diff
@@ -48,36 +50,68 @@ func main() {
+		go func() {
+			// ListenAndServe always returns a non-nil error. After Shutdown or
+			// Close, the returned error is ErrServerClosed
+			if err := s.ListenAndServe(); err != http.ErrServerClosed {
```
@JoelSpeed I've updated this to only log a message if we get an error that isn't `ErrServerClosed` and in that case, we go ahead and close the proxy stop channel as well.

My thinking here is that only the metrics server has encountered an error such as an invalid address, so we want to close that does more gracefully than calling `os.Exit()` via `logger.Fatal()`. I don't think we need to close the metrics channel as the main go routine closes and the process will stop but it sort of seems more correct/complete to do it anyway.

JoelSpeed replied:
> I don't think we need to close the metrics channel as the main go routine closes and the process will stop but it sort of seems more correct/complete to do it anyway.

I'm a bit confused, you say we don't need to but you are closing it below?

So if the start of the metrics server doesn't work, what happens to the main process, does it shutdown gracefully? (If you've tried this do you happen to have the log output?)

NickMeves commented on main.go:
```diff
@@ -53,36 +55,68 @@ func main() {
+
+	// Attempt to setup the metrics endpoint if we have an address
+	if address != "" {
+		s := &http.Server{Addr: address, Handler: middleware.DefaultMetricsHandler}
```
If a user configures TLS for the main server -- I would think we would want the metrics endpoint to listen with the same `tls.Config` right?

JoelSpeed replied:
I think you might actually want to have a second TLS config for the metrics, for example, metrics are likely to be internal so you could just have a self signed cert for metrics for example

For now I might suggest we keep metrics as HTTP only and investigate adding HTTPs support later

neuralsandwich commented:
@JoelSpeed  @NickMeves I'm not sure where we currently are with the review. Is there anything left for me to address? I'm more than happy to of course.

NickMeves commented:
No more feedback from me -- I'll defer to @JoelSpeed for final approval since he took the lead on this PR.

Excited to get metrics over the finish line -- this will be a very nice addition.  Thanks for taking this on!

JoelSpeed commented:
I don't have anything to add now, but there is a conflict, so please fix the conflict up, add a changelog entry and we can get this merged!

neuralsandwich commented:
Added a line to the changelog. Happy to expand on it if needed.

JoelSpeed reviewed (approved):
No, that's all good! Thanks very much, it's been a pleasure working on this with you

neuralsandwich commented:
Uhm, Oops. https://github.com/oauth2-proxy/oauth2-proxy/pull/1049
